lib/Net/Async/Blockchain/ETH.pm

```diff
@@ -418,73 +431,77 @@ async sub _set_transaction_type {
-        my $address = substr($transaction->data, 10, 64);
-        my $amount  = substr($transaction->data, 74, 64);
-
-        return undef unless $address && $amount;
-
-        my $contract_address = $transaction->to;
-
-        my $hex_symbol = await $self->rpc_client->call({
-                data => SYMBOL_SIGNATURE,
-                to   => $contract_address
-            },
-            "latest"
-        );
+    my $logs = $receipt->{logs};
 
-        my $symbol = $self->_to_string($hex_symbol);
-        return undef unless $symbol;
+    return () unless $logs && @$logs;
 
-        my $decimals = await $self->rpc_client->call({
-                data => DECIMALS_SIGNATURE,
-                to   => $contract_address
-            },
-            "latest"
-        );
+    my @transactions;
 
-        if ($decimals) {
-            $transaction->{amount} = Math::BigFloat->from_hex($amount)->bdiv(Math::BigInt->new(10)->bpow($decimals))->bround(hex $decimals);
-        } else {
-            $transaction->{amount} = Math::BigFloat->from_hex($amount);
-        }
+    return () unless $receipt->{status} && hex($receipt->{status}) == 1;
+
+    for my $log ($logs->@*) {
+        my @topics = $log->{topics}->@*;
+        if (@topics && $topics[0] eq TRANSFER_EVENT_SIGNATURE) {
+            my $transaction_cp = $transaction->clone();
+
+            my $address = $log->{address};
+            my $amount  = $self->get_numeric_from_hex($log->{data});
+            next unless $amount;
+
+            my $hex_symbol = await $self->rpc_client->call({
+                    data => SYMBOL_SIGNATURE,
+                    to   => $address
+                },
+                "latest"
+            );
+
+            my $symbol = $self->_to_string($hex_symbol);
+            next unless $symbol;
+
+            $transaction_cp->{currency} = $symbol;
 
-        $transaction->{currency} = $symbol;
-        $transaction->{contract} = $contract_address;
-        $transaction->{to}       = $self->_remove_zeros($address);
+            my $decimals = await $self->rpc_client->call({
+                    data => DECIMALS_SIGNATURE,
+                    to   => $address
+                },
+                "latest"
+            );
 
-        return $transaction;
+            if ($decimals) {
+                my $bg_decimals = $self->get_numeric_from_hex($decimals);
+                # decimals can be 0 is that why we check by reference
+                next unless eval { $bg_decimals->isa('Math::BigFloat') };
+                $transaction_cp->{amount} = $amount->bdiv(Math::BigInt->new(10)->bpow($bg_decimals));
+            } else {
+                $transaction_cp->{amount} = $amount;
+            }
+
+            if (@topics > 1) {
+                $transaction_cp->{to} = $self->_remove_zeros($topics[2]);
+            }
+
+            $transaction_cp->{contract} = $address;
+
+            push(@transactions, $transaction_cp);
```

Comment:
Stylistically it is unusual to put parens around core perl ops like `push`. Usually it would be

    push @transactions, $transaction_cp;